File: core/src/test/java/de/qabel/core/repository/AbstractSqliteRepositoryTest.java
package de.qabel.core.repository;

import de.qabel.core.repository.sqlite.ClientDatabase;
import de.qabel.core.repository.sqlite.DesktopClientDatabase;
import org.junit.After;
import org.junit.Before;

import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.Statement;

public abstract class AbstractSqliteRepositoryTest<T> {
    protected Connection connection;
    protected ClientDatabase clientDatabase;
    protected Path dbFile;
    protected T repo;
    protected EntityManager em;

    @Before
    public void setUp() throws Exception {
        connection = DriverManager.getConnection("jdbc:sqlite::memory:");
        clientDatabase = createDatabase(connection);
        clientDatabase.migrate();
        em = new EntityManager();
        repo = createRepo(clientDatabase, em);
    }

    protected ClientDatabase createDatabase(Connection connection) {
        return new DesktopClientDatabase(connection);
    }

    protected abstract T createRepo(ClientDatabase clientDatabase, EntityManager em) throws Exception;

    @After
    public void tearDown() throws Exception {
        connection.close();
    }
}

File: core/src/test/java/de/qabel/core/repository/AbstractSqliteRepositoryTest.kt
package de.qabel.core.repository

import de.qabel.core.repository.EntityManager
import de.qabel.core.repository.sqlite.ClientDatabase
import de.qabel.core.repository.sqlite.DesktopClientDatabase
import de.qabel.core.repository.sqlite.use
import org.junit.After
import org.junit.Before

import java.nio.file.Path
import java.sql.Connection
import java.sql.DriverManager
import java.sql.Statement

abstract class AbstractSqliteRepositoryTest<T: Any> {
    lateinit protected var connection: Connection
    lateinit protected var clientDatabase: ClientDatabase
    lateinit protected var dbFile: Path
    lateinit protected var repo: T
    lateinit protected var em: EntityManager

    @Before
    @Throws(Exception::class)
    open fun setUp() {
        connection = DriverManager.getConnection("jdbc:sqlite::memory:")
        connection.createStatement().use { statement -> statement.execute("PRAGMA FOREIGN_KEYS = ON") }
        clientDatabase = DesktopClientDatabase(connection)
        clientDatabase.migrate()
        em = EntityManager()
        repo = createRepo(clientDatabase, em)
    }

    @Throws(Exception::class)
    protected abstract fun createRepo(clientDatabase: ClientDatabase, em: EntityManager): T

    @After
    @Throws(Exception::class)
    fun tearDown() {
        connection.close()
    }
}
